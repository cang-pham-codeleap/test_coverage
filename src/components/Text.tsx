import React from "react";

/**
 * Props for the Text component
 * @property {React.ReactNode} children - The content to be displayed
 * @property {React.ElementType} [as='p'] - The HTML element to render as
 * @property {'xs' | 'sm' | 'base' | 'lg' | 'xl' | '2xl'} [size='base'] - Text size
 * @property {'normal' | 'medium' | 'semibold' | 'bold'} [weight='normal'] - Font weight
 * @property {'default' | 'muted' | 'error' | 'success'} [variant='default'] - Text color variant
 * @property {string} [className] - Additional CSS classes
 * @property {boolean} [truncate=false] - Whether to truncate text with ellipsis
 */
export interface TextProps {
  children: React.ReactNode;
  as?: React.ElementType;
  size?: "xs" | "sm" | "base" | "lg" | "xl" | "2xl";
  weight?: "normal" | "medium" | "semibold" | "bold";
  variant?: "default" | "muted" | "error" | "success";
  className?: string;
  truncate?: boolean;
  [key: string]: any; // For additional HTML attributes
}

/**
 * Text component for displaying content with various styles
 *
 * @example
 * ```tsx
 * <Text>Default paragraph text</Text>
 * <Text as="h1" size="2xl" weight="bold">Heading</Text>
 * <Text variant="muted" size="sm">Small muted text</Text>
 * <Text variant="error">Error message</Text>
 * <Text truncate>This is a very long text that will be truncated...</Text>
 * ```
 */
export const Text: React.FC<TextProps> = ({
  children,
  as: Component = "p",
  size = "base",
  weight = "normal",
  variant = "default",
  className = "",
  truncate = false,
  ...props
}) => {
  // Size classes
  const sizeClasses = {
    xs: "text-xs",
    sm: "text-sm",
    base: "text-base",
    lg: "text-lg",
    xl: "text-xl",
    "2xl": "text-2xl",
  };

  // Weight classes
  const weightClasses = {
    normal: "font-normal",
    medium: "font-medium",
    semibold: "font-semibold",
    bold: "font-bold",
  };

  // Variant (color) classes
  const variantClasses = {
    default: "text-gray-900",
    muted: "text-gray-500",
    error: "text-red-600",
    success: "text-green-600",
  };

  // Create CSS classes based on props
  const classes = [
    sizeClasses[size],
    weightClasses[weight],
    variantClasses[variant],
    truncate ? "truncate" : "",
    className,
  ]
    .filter(Boolean)
    .join(" ");

  return (
    <Component className={classes} {...props}>
      {children}
    </Component>
  );
};

export default Text;
