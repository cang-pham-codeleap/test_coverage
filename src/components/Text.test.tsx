/**
 * Text component tests
 * - Tests basic rendering
 * - Tests different sizes, weights and variants
 * - Tests truncate property
 * - Tests polymorphic "as" prop
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import Text from "./Text";

describe("Text Component", () => {
  // Test basic rendering
  test("renders text with children", () => {
    render(<Text>Sample Text</Text>);
    const textElement = screen.getByText("Sample Text");
    expect(textElement).toBeInTheDocument();
    expect(textElement).toMatchSnapshot();
  });

  // Test default rendering as paragraph
  test("renders as paragraph by default", () => {
    render(<Text>Default Text</Text>);
    const textElement = screen.getByText("Default Text");
    expect(textElement.tagName).toBe("P");
  });

  // Test polymorphic "as" prop
  test("renders as specified element when 'as' prop is provided", () => {
    render(<Text as="h1">Heading Text</Text>);
    const textElement = screen.getByText("Heading Text");
    expect(textElement.tagName).toBe("H1");
  });

  // Test different sizes
  test("applies correct size classes", () => {
    const { rerender } = render(<Text size="xs">Extra Small</Text>);
    let textElement = screen.getByText("Extra Small");
    expect(textElement.className).toContain("text-xs");

    rerender(<Text size="lg">Large Text</Text>);
    textElement = screen.getByText("Large Text");
    expect(textElement.className).toContain("text-lg");

    rerender(<Text size="2xl">Extra Large</Text>);
    textElement = screen.getByText("Extra Large");
    expect(textElement.className).toContain("text-2xl");
  });

  // Test different weights
  test("applies correct weight classes", () => {
    const { rerender } = render(<Text weight="normal">Normal Weight</Text>);
    let textElement = screen.getByText("Normal Weight");
    expect(textElement.className).toContain("font-normal");

    rerender(<Text weight="medium">Medium Weight</Text>);
    textElement = screen.getByText("Medium Weight");
    expect(textElement.className).toContain("font-medium");

    rerender(<Text weight="bold">Bold Weight</Text>);
    textElement = screen.getByText("Bold Weight");
    expect(textElement.className).toContain("font-bold");
  });

  // Test different variants
  test("applies correct variant classes", () => {
    const { rerender } = render(<Text variant="default">Default Variant</Text>);
    let textElement = screen.getByText("Default Variant");
    expect(textElement.className).toContain("text-gray-900");

    rerender(<Text variant="muted">Muted Variant</Text>);
    textElement = screen.getByText("Muted Variant");
    expect(textElement.className).toContain("text-gray-500");

    rerender(<Text variant="error">Error Variant</Text>);
    textElement = screen.getByText("Error Variant");
    expect(textElement.className).toContain("text-red-600");

    rerender(<Text variant="success">Success Variant</Text>);
    textElement = screen.getByText("Success Variant");
    expect(textElement.className).toContain("text-green-600");
  });

  // Test truncate property
  test("applies truncate class when truncate is true", () => {
    render(<Text truncate>Truncated Text</Text>);
    const textElement = screen.getByText("Truncated Text");
    expect(textElement.className).toContain("truncate");
  });

  // Test custom class name
  test("applies custom className when provided", () => {
    const customClass = "custom-text-class";
    render(<Text className={customClass}>Custom Style</Text>);
    const textElement = screen.getByText("Custom Style");
    expect(textElement.className).toContain(customClass);
  });

  // Test additional attributes
  test("applies additional HTML attributes", () => {
    render(<Text data-testid="custom-text" id="special-text">Custom Attribute</Text>);
    const textElement = screen.getByTestId("custom-text");
    expect(textElement).toHaveAttribute("id", "special-text");
    expect(textElement).toMatchSnapshot();
  });
});
