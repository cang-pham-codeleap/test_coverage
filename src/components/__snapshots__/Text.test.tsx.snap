// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component applies additional HTML attributes 1`] = `
<p
  class="text-base font-normal text-gray-900"
  data-testid="custom-text"
  id="special-text"
>
  Custom Attribute
</p>
`;

exports[`Text Component renders text with children 1`] = `
<p
  class="text-base font-normal text-gray-900"
>
  Sample Text
</p>
`;
